python/ambassador/config/resourceprocessor.py

```diff
@@ -0,0 +1,671 @@
+from __future__ import annotations
+from typing import Any, ClassVar, ContextManager, Dict, FrozenSet, Iterator, List, Mapping, Optional, Type, TypeVar, Set
+
+import collections
+import collections.abc
+import contextlib
+import dataclasses
+import datetime
+import itertools
+import json
+import logging
+
+import durationpy
+
+from .config import Config
+from .acresource import ACResource
+
+from ..utils import dump_yaml
+
+
+@dataclasses.dataclass
+class Location:
+    """
+    Represents a location for parsing.
+    """
+
+    filename: Optional[str] = None
+    ocount: int = 1
+
+    def mark_annotation(self) -> None:
+        if self.filename is None:
+            return
+        elif self.filename.endswith(':annotation'):
+            return
+
+        self.filename += ':annotation'
+
+    def __str__(self) -> str:
+        return f"{self.filename or 'anonymous YAML'}.{self.ocount}"
+
+
+class LocationManager:
+    """
+    Manages locations contextually.
+    """
+
+    previous: List[Location]
+    current: Location
+
+    def __init__(self) -> None:
+        self.previous = []
+        self.current = Location()
+
+    def push(self, filename: Optional[str] = None, ocount: int = 1) -> ContextManager[Location]:
+        current = Location(filename, ocount)
+        self.previous.append(self.current)
+        self.current = current
+
+        # This trick lets you use the return value of this method in a `with`
+        # statement. At the conclusion of the statement block, the location will
+        # automatically be popped from the stack.
+        @contextlib.contextmanager
+        def popper():
+            yield current
+            self.pop()
+
+        return popper()
+
+    def push_reset(self) -> ContextManager[Location]:
```

Comment:
I wasn't 100% sure if the justification for this method existing is correct but it's how parse_object works when used inside the Kubernetes handler. Happy to accept suggestions for a better name/docstring.